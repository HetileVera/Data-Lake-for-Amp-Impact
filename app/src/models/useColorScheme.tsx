import { useTheme } from "@chakra-ui/react";

export function useColorScheme() {
  const colorScheme = useTheme().name;
  const tone = (value: number | undefined) => (value ? `${colorScheme}.${value}` : colorScheme);

  return { tone, colorScheme };
}
