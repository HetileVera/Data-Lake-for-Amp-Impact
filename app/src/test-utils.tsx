import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";

import { theme } from "themes/orange";

export const CustomChakraProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <ChakraProvider theme={theme}>
      <div data-testid="test-root">{children}</div>
    </ChakraProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "wrapper">) => render(ui, { wrapper: CustomChakraProvider, ...options });

export * from "@testing-library/react";

export { customRender as render };
