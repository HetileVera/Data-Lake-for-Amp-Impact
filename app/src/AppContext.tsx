import React, { createContext, useContext } from "react";
import isEmpty from "lodash/isEmpty";
import throttle from "lodash/throttle";

import { AppStore, IAppStore } from "./models/AppStore";
import { IInstallation } from "./models/Installation";
import { isLocalDev, websiteUrlOrigin, assetWebsiteUrl, secretKey } from "./helpers/settings";
import { onSnapshot } from "mobx-state-tree";

// Create an instance of the root store (a.k.a appStore)
const appStore = AppStore.create({});

// Store state to local storage
const storeLocally = throttle(() => appStore.saveLocal(), 5000);

// Store state in S3 when applicable
const storeRemotely = throttle(() => appStore.saveRemote(), 30000);

// The initialization function for the root store
export function initializeStore() {
  if (!isEmpty(secretKey)) console.warn("Access key is bundled!");

  if (isLocalDev) {
    console.log(`isLocalDev '${isLocalDev}'`);
    console.log(`websiteUrlOrigin '${websiteUrlOrigin}'`);
    console.log(`assetWebsiteUrl '${assetWebsiteUrl}'`);

    if (window.location.host !== "localhost:3000") {
      console.warn("You should run the app on localhost!");

      throw new Error("App should run on localhost for development mode!");
    }
  }

  appStore.loadLocal();
  onSnapshot(appStore, (snapshot) => {
    storeLocally();
    storeRemotely();
  });
}

// Create the store context
export const StoreContext = createContext<IAppStore | undefined>(undefined);

// Create the store provider component
export const StoreProvider: React.FunctionComponent = ({ children }) => {
  return <StoreContext.Provider value={appStore}>{children}</StoreContext.Provider>;
};

// useStore hook
export function useStore(): IAppStore {
  const store = useContext(StoreContext);

  if (!store) {
    throw new Error("useStore() must be used within a StoreProvider");
  }

  return store;
}

// useInstallation hook
export function useInstallation(): IInstallation {
  const store = useStore();
  const installation = store.installation;

  if (!installation) {
    throw new Error("An installation instance was not created yet");
  }

  return installation;
}
