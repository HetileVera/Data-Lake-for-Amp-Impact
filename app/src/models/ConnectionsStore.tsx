import { useEffect } from "react";
import isEmpty from "lodash/isEmpty";
import { types, Instance, getParent } from "mobx-state-tree";

import { useInstallation } from "AppContext";
import { listAppflowConnections } from "api/list-appflow-connections";

import { BaseStore, isStoreNew, isStoreError, isStoreReady, isStoreLoading, isStoreReLoading } from "./BaseStore";
import { ICredentials } from "./helpers/Credentials";
import { IInstallation } from "./Installation";

/**
 * Represents the store for the AppFlow Salesforce connection names
 */
export const ConnectionsStore = BaseStore.named("ConnectionsStore")
  .props({
    connectionNames: types.array(types.string),
  })

  .views((self) => ({
    get empty() {
      return isEmpty(self.connectionNames);
    },

    get credentials(): ICredentials {
      const parent: IInstallation = getParent(self);
      return parent.credentials;
    },

    get region(): string {
      const parent: IInstallation = getParent(self);
      return parent.region;
    },
  }))

  .actions((self) => {
    const superReset = self.reset;
    return {
      reset() {
        superReset();
        self.connectionNames.clear();
      },

      async doLoad() {
        const accessKey = self.credentials.accessKey;
        const secretKey = self.credentials.secretKey;
        const region = self.region;

        const rawConnectionNames = await listAppflowConnections({ accessKey, secretKey, region });
        self.runInAction(() => {
          self.connectionNames.replace(rawConnectionNames);
        });
      },
    };
  });

// see https://mobx-state-tree.js.org/tips/typescript
export interface IConnectionsStore extends Instance<typeof ConnectionsStore> {}

export function useConnectionsStore() {
  const installation = useInstallation();
  const store = installation.connectionsStore;

  useEffect(() => {
    if (!isStoreNew(store)) return;
    store.load();
  }, [store]);

  return {
    isError: isStoreError(store),
    isReady: isStoreReady(store),
    isLoading: isStoreLoading(store),
    isReloading: isStoreReLoading(store),
    store,
  };
}
