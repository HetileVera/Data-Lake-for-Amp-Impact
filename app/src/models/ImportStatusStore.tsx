import includes from "lodash/includes";
import split from "lodash/split";
import last from "lodash/last";
import chunkIt from "lodash/chunk";
import first from "lodash/first";
import isEmpty from "lodash/isEmpty";
import sortBy from "lodash/sortBy";
import { ExecutionStatus, HistoryEvent } from "@aws-sdk/client-sfn";
import { types, Instance, getParent } from "mobx-state-tree";
import { values } from "mobx";
import { useEffect } from "react";

import { useInstallation } from "AppContext";
import { s3GetJson } from "api/s3-get-json";
import { describeSFN } from "api/describe-sfn";
import { getExecutionFailureInfo, getSFNHistory, isExecutionFailureEvent } from "api/get-sfn-history";
import { s3ListObjects, S3Object } from "api/list-s3-objects";
import { getLatestSFN } from "api/get-lastest-sfn";
import { delay, niceByte, niceNumber } from "helpers/utils";

import { BaseStore, isStoreError, isStoreLoading, isStoreNew, isStoreReady, isStoreReLoading } from "./BaseStore";
import { ICredentials } from "./helpers/Credentials";
import { IInstallation } from "./Installation";

const isSuccessStatus = (status: ExecutionStatus) => {
  // For all statuses, see https://docs.aws.amazon.com/step-functions/latest/apireference/API_DescribeExecution.html

  // We consider any of these statuses as a success
  const successStatuses = ["SUCCEEDED"];
  return includes(successStatuses, status);
};

const isFailureStatus = (status: ExecutionStatus) => {
  // For all statuses, see https://docs.aws.amazon.com/step-functions/latest/apireference/API_DescribeExecution.html

  // We consider any of these statuses as a failure
  const failureStatuses = ["FAILED", "ABORTED", "TIMED_OUT"];
  return includes(failureStatuses, status);
};

export enum ObjectImportStatusCode {
  Not_Started = "NOT_STARTED",
  Preparing = "PREPARING",
  In_Progress = "IN_PROGRESS",
  Importing = "IMPORTING",
  Successful = "SUCCESSFUL",
}

interface ObjectStatusInput {
  name: string;
  label?: string;
  status?: ObjectImportStatusCode;
  rowCount?: number;
  columnCount?: number;
  importSeconds?: number;
  bytesWritten?: number;
}

interface RemoteObject {
  status: ObjectImportStatusCode;
  metadata?: {
    rowCount: number;
    columnCount: number;
    importSeconds: number;
    bytesWritten: number;
  };
}

/**
 * Represents a Salesforce object import status
 */
const ObjectImportStatus = types
  .model("ObjectImportStatus", {
    name: "",
    label: "",
    status: types.optional(
      types.enumeration<ObjectImportStatusCode>("ObjectImportStatusCode", Object.values(ObjectImportStatusCode)),
      ObjectImportStatusCode.Not_Started
    ),
    rowCount: 0,
    columnCount: 0,
    importSeconds: 0,
    bytesWritten: 0,
    stale: false,
  })

  .actions((self) => ({
    setObjectStatus({ name, label, status, rowCount, columnCount, importSeconds, bytesWritten }: ObjectStatusInput) {
      self.name = name;
      if (label) self.label = label;
      if (status) self.status = status;
      if (rowCount) self.rowCount = rowCount;
      if (columnCount) self.columnCount = columnCount;
      if (importSeconds) self.importSeconds = importSeconds;
      if (bytesWritten) self.bytesWritten = bytesWritten;
    },

    setStale(flag: boolean) {
      self.stale = flag;
      if (flag) self.status = ObjectImportStatusCode.Not_Started;
    },
  }))

  .views((self) => ({
    get isNotStarted(): boolean {
      return self.status === ObjectImportStatusCode.Not_Started;
    },

    get isPreparing(): boolean {
      return self.status === ObjectImportStatusCode.Preparing;
    },

    get isInProgress(): boolean {
      return self.status === ObjectImportStatusCode.In_Progress;
    },

    get isImporting(): boolean {
      return self.status === ObjectImportStatusCode.Importing;
    },

    get isSuccessful(): boolean {
      return self.status === ObjectImportStatusCode.Successful;
    },

    get warning(): boolean {
      return self.rowCount < 0;
    },
  }))

  .views((self) => ({
    get statsAvailable(): boolean {
      // Tells us if any of the stats such as (rowCount, et) are available.
      return self.isSuccessful;
    },
  }))

  .views((self) => ({
    get rowCountDisplay(): string {
      if (!self.statsAvailable) return "N/A";

      return niceNumber(self.rowCount);
    },

    get columnCountDisplay(): string {
      if (!self.statsAvailable) return "N/A";

      return niceNumber(self.columnCount);
    },

    get importSecondsDisplay(): string {
      if (!self.statsAvailable) return "N/A";

      return `${Math.round(self.importSeconds)} sec`;
    },

    get bytesWrittenDisplay(): string {
      if (!self.statsAvailable) return "N/A";

      return niceByte(self.bytesWritten);
    },
  }));

export interface IObjectImportStatus extends Instance<typeof ObjectImportStatus> {}

export enum LoadingStage {
  GettingSchemas = "GETTING_SCHEMAS",
  MonitoringProgress = "MONITORING_PROGRESS",
}

/**
 * Represents the model for the status of importing salesforce objects. This includes each object import status.
 */
export const ImportStatusStore = BaseStore.named("ImportStatusStore")
  .props({
    objects: types.map(ObjectImportStatus),
    startTime: 0, // The start time of the execution (epoch milliseconds)
    loadingStage: types.maybe(types.enumeration<LoadingStage>("LoadingStage", Object.values(LoadingStage))),
    _errorDetail: "", // Expected to be in markdown https://commonmark.org/help/
  })

  .views((self) => ({
    get empty() {
      return self.objects.size === 0;
    },

    get credentials(): ICredentials {
      const parent: IInstallation = getParent(self);
      return parent.credentials;
    },

    get region(): string {
      const parent: IInstallation = getParent(self);
      return parent.region;
    },

    get importWorkflowArn(): string {
      const parent: IInstallation = getParent(self);
      return parent.importWorkflowArn;
    },

    get metadataBucket(): string {
      const parent: IInstallation = getParent(self);
      return parent.metadataBucket;
    },

    get stateMachineArn(): string {
      const parent: IInstallation = getParent(self);
      return parent.importWorkflowArn;
    },

    get isGettingSchemas() {
      return self.loadingStage === LoadingStage.GettingSchemas && (isStoreLoading(self) || isStoreReLoading(self));
    },

    get isMonitoringProgress() {
      return self.loadingStage === LoadingStage.MonitoringProgress && (isStoreLoading(self) || isStoreReLoading(self));
    },

    get errorDetail() {
      return self._errorDetail;
    },

    get listObjects(): IObjectImportStatus[] {
      const result: readonly unknown[] = values(self.objects);
      return sortBy(result as IObjectImportStatus[], ["label", "name"]);
    },

    get objectsCount(): number {
      return self.objects.size;
    },

    get processPercentage(): number {
      const size = self.objects.size;
      let count = 0;

      self.objects.forEach((object) => {
        switch (object.status) {
          case ObjectImportStatusCode.Preparing:
            count += 0.25;
            break;
          case ObjectImportStatusCode.In_Progress:
            count += 0.5;
            break;
          case ObjectImportStatusCode.Importing:
            count += 0.75;
            break;
          case ObjectImportStatusCode.Successful:
            count += 1;
        }
      });

      if (size === 0) return 0; // Avoid division by zero
      return (count / size) * 100;
    },
  }))

  .views((self) => ({
    get isSuccess(): boolean {
      return isStoreReady(self) && !isStoreError(self) && isEmpty(self.errorDetail);
    },

    get totalRecords(): number {
      return self.listObjects.reduce<number>((result, item) => result + Math.max((item as IObjectImportStatus).rowCount, 0), 0);
    },

    get totalFields(): number {
      return self.listObjects.reduce<number>((result, item) => result + Math.max((item as IObjectImportStatus).columnCount, 0), 0);
    },

    get totalBytes(): number {
      return self.listObjects.reduce<number>((result, item) => result + Math.max((item as IObjectImportStatus).bytesWritten, 0), 0);
    },

    get totalTime(): number {
      return self.listObjects.reduce<number>((result, item) => result + Math.max((item as IObjectImportStatus).importSeconds, 0), 0);
    },
  }))

  .views((self) => ({
    get totalRecordsDisplay(): string {
      return niceNumber(self.totalRecords);
    },

    get totalFieldsDisplay(): string {
      return niceNumber(self.totalFields);
    },

    get totalBytesDisplay(): string {
      return niceByte(self.totalBytes);
    },

    get totalTimeDisplay(): string {
      return `${Math.round(self.totalTime)} sec`;
    },
  }))

  .actions((self) => ({
    setObject(rawObject: ObjectStatusInput) {
      if (self.objects.has(rawObject.name)) {
        self.objects.get(rawObject.name)?.setObjectStatus(rawObject);
      } else {
        self.objects.set(rawObject.name, rawObject);
      }

      return self.objects.get(rawObject.name);
    },

    setLoadingStage(stage: LoadingStage) {
      self.loadingStage = stage;
    },

    clearLoadingStage() {
      self.loadingStage = undefined;
    },

    setStartTime(time: number) {
      self.startTime = time;
    },
  }))

  .actions((self) => ({
    buildStateMachineUrl(): string {
      const region = self.region;
      const stateMachineArn = self.importWorkflowArn;

      return `https://${region}.console.aws.amazon.com/states/home?region=${region}#/statemachines/view/${encodeURIComponent(stateMachineArn)}`;
    },

    buildExecutionUrl(executionArn: string): string {
      const region = self.region;

      return `https://${region}.console.aws.amazon.com/states/home?region=${region}#/executions/details/${encodeURIComponent(executionArn)}`;
    },
  }))

  .actions((self) => ({
    async describeWorkflow(executionArn: string): Promise<ExecutionStatus> {
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const workflowArn = self.importWorkflowArn;

      try {
        const info = await describeSFN({ accessKey, secretKey, region, executionArn });
        return info.status as ExecutionStatus;
      } catch (error) {
        self.runInAction(() => {
          // This is in markdown
          self._errorDetail = `We attempted to check the status of the execution of the state machine by calling the Step Functions DescribeExecution API.
          However, we got the following error:
### ${error}

### Information
* State machine execution ARN: ${executionArn}
* State machine ARN: ${workflowArn}

### Links
* [AWS Step Functions execution page](${self.buildExecutionUrl(executionArn)})
* [AWS Step Functions state machine page](${self.buildStateMachineUrl()})
* [Step Functions DescribeExecution API documentation](https://docs.aws.amazon.com/step-functions/latest/apireference/API_DescribeExecution.html)
          `;
        });
        throw error;
      }
    },

    async handleSFNFailure(executionArn: string, status: ExecutionStatus) {
      // In this method we display an informative error detail that provides enough context around the failure.
      // To do this, we will list all the execution history events and look for the overall execution event that had an error.
      // Since getting the history events is not critical, we will capture any errors from calling GetExecutionHistory
      // and not propagate them or have them affecting the overall status of the operation.

      let executionEvent: HistoryEvent;
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const workflowArn = self.importWorkflowArn;

      try {
        const events = await getSFNHistory({ accessKey, secretKey, region, executionArn });
        // We find the execution event that failed
        for (const event of events) {
          if (isExecutionFailureEvent(event)) {
            // We found the event
            executionEvent = event;
            break;
          }
        }
      } catch (error) {
        // Since getting the stack events is not critical, we will capture any errors from calling the describe stack events
        // call and not propagate them or have them affect the overall status of the operation.
        console.log(error);
      }

      self.runInAction(() => {
        // This is in markdown
        self._errorDetail = `The step function workflow failed with the following status:
### ${status}`;

        if (executionEvent) {
          const info = getExecutionFailureInfo(executionEvent);
          self._errorDetail = `${self._errorDetail}
### Execution Failure
* Cause: ${info.cause}
         `;
        }

        self._errorDetail = `${self._errorDetail}
### Information
* State machine execution ARN: ${executionArn}
* State machine ARN: ${workflowArn}
        
### Links
* [AWS Step Functions execution page](${self.buildExecutionUrl(executionArn)})
* [AWS Step Functions state machine page](${self.buildStateMachineUrl()})        
          `;
      });
      throw new Error(status);
    },

    async loadSchemas() {
      // We need to detect all the objects that we are expecting to import, this is done by looking at the schema files in S3
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const bucketName = self.metadataBucket;
      const worker = async (s3Object: S3Object) => {
        const json = (await s3GetJson({ accessKey, secretKey, region, bucketName, objectKey: s3Object.key })) as { name: string; label: string };
        self.setObject({ name: json.name, label: json.label ?? json.name });
      };

      const objects = await s3ListObjects({ accessKey, secretKey, region, bucketName, prefix: `schemas/` });
      const chunks = chunkIt(objects, 5); // chunks them 5 at a time
      while (!isEmpty(chunks)) {
        const chunk = chunks.shift() ?? [];
        // Create a list of worker promises and wait on all of them
        await Promise.all(chunk.map((object) => worker(object)));
      }
    },

    async pullProgress(executionArn: string) {
      // We pull that latest status of any object that has not reached the successful stage
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const runId = last(split(executionArn, ":"));
      const bucketName = self.metadataBucket;
      const worker = async (s3Object: S3Object) => {
        // An example of an S3 object key is
        // sf-metadata-u1puzajv0crm/state/runs/aaaaa-8d51-4dd3-bbf7-a81ee93cdc9a/Account.status.json
        // So, can split on "/" and take the last element which is "Account.status.json" and then
        // split on "." and take the first element which give us "Account"
        const name = first(split(last(split(s3Object.key, "/")), ".")) ?? ""; // Yes, we can regex but this is also okay, see explanation above
        try {
          const existing = self.objects.get(name);
          if (existing?.isSuccessful) return;

          const json = (await s3GetJson({ accessKey, secretKey, region, bucketName, objectKey: s3Object.key })) as RemoteObject;
          const object = self.setObject({
            name,
            status: json.status,
            rowCount: json.metadata?.rowCount,
            columnCount: json.metadata?.columnCount,
            importSeconds: json.metadata?.importSeconds,
            bytesWritten: json.metadata?.bytesWritten,
          });

          object?.setStale(false);
        } catch (error) {
          // NOTE: it is important that we don't propagate this error, we don't want status reporting to halt deployment
          console.error(error);
        }
      };

      const objects = await s3ListObjects({ accessKey, secretKey, region, bucketName, prefix: `state/runs/${runId}/` });
      const chunks = chunkIt(objects, 5); // chunks them 5 at a time
      while (!isEmpty(chunks)) {
        const chunk = chunks.shift() ?? [];
        // Create a list of worker promises and wait on all of them
        await Promise.all(chunk.map((object) => worker(object)));
      }
    },

    async getLatestExecution() {
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const stateMachineArn = self.stateMachineArn;

      return getLatestSFN({ accessKey, secretKey, region, stateMachineArn });
    },

    markStale() {
      self.objects.forEach((object) => {
        object.setStale(true);
      });
    },

    removeStale() {
      self.objects.forEach((object) => {
        if (object.stale) self.objects.delete(object.name); // It is safe to delete from a map while iterating
      });
    },
  }))

  .actions((self) => {
    const superAfterCreate = self.afterCreate;
    const superReset = self.reset;
    return {
      afterCreate() {
        superAfterCreate();
        self._errorDetail = "";
        self.loadingStage = undefined;
      },

      reset() {
        superReset();
        self._errorDetail = "";
        self.objects.clear();
        self.loadingStage = undefined;
      },

      async doLoad({ executionArn, startTime }: { executionArn?: string; startTime?: number } = {}) {
        self.loadingStage = LoadingStage.GettingSchemas;
        self._errorDetail = "";
        self.startTime = startTime ? startTime : self.startTime;
        self.markStale();

        if (!executionArn) {
          const latestExecution = await self.getLatestExecution();
          if (!latestExecution || !latestExecution.executionArn) {
            // If we don't have any, then there is nothing else to do
            self.setStartTime(0);
            self.clearLoadingStage();
            return;
          }
          self.setStartTime(latestExecution.startDate?.getTime() ?? 0);
          executionArn = latestExecution.executionArn;
        }

        // Load the schemas
        await self.loadSchemas();

        let status: ExecutionStatus;
        const shouldStop = (status: ExecutionStatus) => isFailureStatus(status) || isSuccessStatus(status);
        self.setLoadingStage(LoadingStage.MonitoringProgress);

        do {
          status = await self.describeWorkflow(executionArn);
          await delay(10);
          await self.pullProgress(executionArn);
        } while (!shouldStop(status));

        self.clearLoadingStage();
        if (isSuccessStatus(status)) {
          self.removeStale();
          return; // We are done with the success scenario
        }

        // We hit an SFN failure
        await self.handleSFNFailure(executionArn, status);
      },
    };
  });

// see https://mobx-state-tree.js.org/tips/typescript
export interface IImportStatusStore extends Instance<typeof ImportStatusStore> {}

export function useImportStatusStore() {
  const installation = useInstallation();
  const store = installation.importStatusStore;

  useEffect(() => {
    // TODO - add a check for time and if it has been more than 10 minutes, then load the store
    if (!isStoreNew(store)) return;
    store.load();
  }, [store]);

  return { isError: isStoreError(store), store };
}
