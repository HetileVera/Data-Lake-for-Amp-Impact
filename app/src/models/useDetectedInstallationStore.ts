import { isStoreError, isStoreLoading, isStoreReady, isStoreReLoading } from "models/BaseStore";

import { DetectedInstallationStore } from "./DetectedInstallationsStore";

// Create an instance of this store
const store = DetectedInstallationStore.create({});

export function useDetectedInstallationStore() {
  return {
    isError: isStoreError(store),
    isReady: isStoreReady(store),
    isLoading: isStoreLoading(store),
    isReloading: isStoreReLoading(store),
    store,
  };
}
