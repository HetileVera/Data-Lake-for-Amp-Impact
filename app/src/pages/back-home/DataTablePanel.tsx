import React from "react";
import { Box, Heading } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useColorScheme } from "models/useColorScheme";
import { useMetadataStore } from "models/MetadataStore";

import { DataModelTablePreview } from "pages/step4/data-model-table-preview/DataModelTablePreview";

import { CountBadge } from "./CountBadge";
import { DataTableStatusInfo } from "./DataTableStatusInfo";

export const DataTablePanel = observer(() => {
  const { tone, colorScheme } = useColorScheme();
  const { store, isReady } = useMetadataStore();
  const selectedObjectsCount = store.selectedObjects.length;
  const isEmptyModel = selectedObjectsCount === 0;

  return (
    <>
      <Heading size="md" pt="0px" pb="0px" color={tone(600)} letterSpacing="-1px">
        Objects <CountBadge />
      </Heading>
      {isEmptyModel && isReady && (
        <Box bg={tone(100)} textAlign="center" p={4} mt={6} mb={4} fontSize="sm" color={tone(600)} borderRadius="md">
          No objects
        </Box>
      )}
      <DataTableStatusInfo />
      <DataModelTablePreview colorScheme={colorScheme} showCaption={false} />
    </>
  );
});
