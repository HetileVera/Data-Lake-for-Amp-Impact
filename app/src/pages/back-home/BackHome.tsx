import React from "react";
import { Box, Text, Button } from "@chakra-ui/react";
import { ArrowForwardIcon } from "@chakra-ui/icons";
import { observer } from "mobx-react";
import { useNavigate } from "react-router-dom";

import { useInstallation, useStore } from "AppContext";
import { Layout } from "components/Layout";
import { TimeAgo } from "components/TimeAgo";
import { isCompleted } from "models/Installation";
import { useColorScheme } from "models/useColorScheme";
import { AccessInformation } from "pages/step6/AccessInformation";
import { UsersPanel } from "pages/step6/users-panel/UsersPanel";
import { CloudwatchDashboardPanel } from "pages/step6/CloudwatchDashboardPanel";
import { SubscribeSNSPanel } from "pages/step6/SubscribeSNSPanel";
import { DataImportStatusPanel } from "pages/step5/data-import-status-panel/DataImportStatusPanel";

import { ReviewGrid } from "./ReviewGrid";
import { DataTablePanel } from "./DataTablePanel";
import { ManagePanel } from "./ManagePanel";
import { StartNewWarning } from "./StartNewWarning";
import { CredentialsCollectionForm } from "./CredentialsCollectionForm";

export const BackHome = observer(() => {
  const { tone, colorScheme } = useColorScheme();
  const navigate = useNavigate();
  const store = useStore();
  const installation = useInstallation();
  const [collectCredentials, setCollectCredentials] = React.useState(false);
  const [warning, setWarning] = React.useState(false);
  const completedProvisioning = isCompleted(installation.deploymentStep);
  const reviewCompleted = isCompleted(installation.reviewStep);
  const credentialsAvailable = !installation.credentials.empty;

  const gotoNextStep = () => {
    const nextStep = installation.nextStepNumber;
    navigate(`/steps/${nextStep}`);
  };

  const handleCollectCredentials = () => {
    if (credentialsAvailable) return gotoNextStep();
    setCollectCredentials(true);
  };

  const handleStartNew = () => {
    if (reviewCompleted) {
      setWarning(true);
    } else {
      store.newInstallation();
    }
  };

  React.useEffect(() => {
    window.scrollTo(0, 0);
  }, []);

  React.useEffect(() => {
    if (collectCredentials || warning) setTimeout(() => window.scrollBy(0, 300), 200);
  }, [collectCredentials, warning]);

  return (
    <Layout step={0} title="Data Lake Information">
      <Box position="relative" boxShadow="base" borderRadius="lg" bg={tone(50)} p={10} pb={6} color={tone(800)}>
        <Box fontSize="small" mt={1} mb={8} color={tone(700)}>
          <TimeAgo time={installation.startDate} /> &#8226; by {installation.startedBy}
        </Box>

        <ReviewGrid />

        {!completedProvisioning && !collectCredentials && !warning && (
          <>
            <Text pl={2} mt={6}>
              Click <b>Resume</b> to continue the steps of provisioning the data lake
              {!credentialsAvailable && <>, we will ask you for your AWS admin credentials so that we can connect to your AWS account your </>}. You
              can also choose to start a new data lake.
            </Text>
            <Box textAlign="right" w="full" mt={6} mb={0}>
              <Button colorScheme={colorScheme} size="md" variant="outline" mr={6} onClick={handleStartNew}>
                Start New
              </Button>
              <Button colorScheme={colorScheme} size="md" rightIcon={<ArrowForwardIcon />} onClick={handleCollectCredentials}>
                Resume
              </Button>
            </Box>
          </>
        )}
      </Box>

      {completedProvisioning && (
        <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={10}>
          <AccessInformation />
        </Box>
      )}

      {completedProvisioning && credentialsAvailable && (
        <>
          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={4} position="relative">
            <UsersPanel />
          </Box>

          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={6}>
            <DataImportStatusPanel />
          </Box>

          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={6}>
            <DataTablePanel />
          </Box>

          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={4} position="relative">
            <CloudwatchDashboardPanel region={installation.region} id={installation.id} />
          </Box>

          <Box borderRadius="lg" boxShadow="base" bg={tone(25)} mt={6} p={10} pb={4} position="relative">
            <SubscribeSNSPanel />
          </Box>
        </>
      )}

      {completedProvisioning && !credentialsAvailable && <ManagePanel />}

      {warning && <StartNewWarning onCancel={() => setWarning(false)} onContinue={() => store.newInstallation()} />}

      {collectCredentials && (
        <CredentialsCollectionForm
          onCancel={() => setCollectCredentials(false)}
          onDone={() => {
            setCollectCredentials(false);
            gotoNextStep();
          }}
        />
      )}
      <Box h="50px">&nbsp;</Box>
    </Layout>
  );
});
