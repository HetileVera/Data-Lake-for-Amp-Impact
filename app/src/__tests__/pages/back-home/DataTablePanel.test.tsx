import { cleanup, screen } from "@testing-library/react";
import { create } from "react-test-renderer";

import { render, RenderWithChakra } from "test-utils";

import { StoreProvider } from "AppContext";

import { DataTablePanel } from "pages/back-home/DataTablePanel";

import * as metadataStore from "models/MetadataStore";

describe("BackHome -> DataTablePanel component", () => {
  afterEach(cleanup);

  test("snapshot", () => {
    // @ts-ignore
    jest.spyOn(metadataStore, "useMetadataStore").mockImplementation(() => {
      return {
        isError: false,
        isReady: true,
        store: { selectedObjects: [{}] },
      };
    });

    const tree = create(
      <RenderWithChakra>
        <StoreProvider>
          <DataTablePanel />
        </StoreProvider>
      </RenderWithChakra>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("render properly with 2 selected objects", () => {
    // @ts-ignore
    jest.spyOn(metadataStore, "useMetadataStore").mockImplementation(() => {
      return {
        isError: false,
        isReady: true,
        store: { selectedObjects: [{}, {}] },
      };
    });

    render(
      <StoreProvider>
        <DataTablePanel />
      </StoreProvider>
    );

    expect(screen.getByText("Objects")).toBeInTheDocument();
    expect(screen.getByText("2")).toBeInTheDocument();
  });

  test("render properly with no selected objects", () => {
    // @ts-ignore
    jest.spyOn(metadataStore, "useMetadataStore").mockImplementation(() => {
      return {
        isError: false,
        isReady: true,
        store: { selectedObjects: [] },
      };
    });

    render(
      <StoreProvider>
        <DataTablePanel />
      </StoreProvider>
    );

    expect(screen.getByText("No objects")).toBeInTheDocument();
  });
});
