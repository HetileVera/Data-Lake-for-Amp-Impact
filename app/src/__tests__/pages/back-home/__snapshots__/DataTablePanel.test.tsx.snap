// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackHome -> DataTablePanel component snapshot 1`] = `
Array [
  <h2
    className="chakra-heading css-z71ux5"
  >
    Objects 
    <span
      className="chakra-badge css-1hhvihq"
    >
      1
    </span>
  </h2>,
  <div
    className="css-8hr4wd"
  >
    <div
      className="css-1lpnqa0"
    >
      Retrieving the data schema, this might take a few seconds
      <br />
      NaN
       %
    </div>
    <div
      className="css-1jer4t"
    >
      <div
        aria-valuemax={100}
        aria-valuemin={0}
        className="css-1585z0b"
        data-indeterminate=""
        role="progressbar"
        style={
          Object {
            "width": "0%",
          }
        }
      />
    </div>
  </div>,
  <table
    className="chakra-table css-94ssnp"
    role="table"
  >
    <thead
      className="css-0"
    >
      <tr
        className="css-0"
        role="row"
      >
        <th
          className="css-y40s86"
        >
          Type
        </th>
        <th
          className="css-y40s86"
          data-is-numeric={true}
        >
          # Fields
        </th>
      </tr>
    </thead>
    <tbody
      className="css-0"
    >
      <tr
        className="css-4g6ai3"
        onClick={[Function]}
        role="row"
      >
        <td
          className="css-1etw0vg"
          role="gridcell"
        >
          <div
            className="chakra-stack css-snbr68"
          >
            <svg
              className="chakra-icon css-m6c8f6"
              focusable={false}
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                fill="currentColor"
              />
            </svg>
            <div
              className="css-0"
            >
               Â 
              <p
                className="chakra-text css-1fk7407"
              >
                ( 
                 )
              </p>
            </div>
          </div>
        </td>
        <td
          className="css-1etw0vg"
          data-is-numeric={true}
          role="gridcell"
        >
          N/A
          /
          N/A
        </td>
      </tr>
    </tbody>
    <tfoot
      className="css-5wov5z"
    >
      <tr
        className="css-0"
        role="row"
      >
        <th
          className="css-y40s86"
        />
        <th
          className="css-y40s86"
          data-is-numeric={true}
        >
          N/A
        </th>
      </tr>
    </tfoot>
  </table>,
  <span />,
]
`;
